src/types/proof/definition.rs

```diff
@@ -187,28 +192,282 @@ impl<H: SimpleHasher> SparseMerkleProof<H> {
+                    SparseMerkleInternalNode::new(sibling_node.hash::<H>(), hash).hash::<H>()
                 } else {
-                    SparseMerkleInternalNode::<H>::new(hash, *sibling_hash).hash()
+                    SparseMerkleInternalNode::new(hash, sibling_node.hash::<H>()).hash::<H>()
                 }
             });
+
         ensure!(
             actual_root_hash == expected_root_hash.0,
             "Root hashes do not match. Actual root hash: {:?}. Expected root hash: {:?}.",
             actual_root_hash,
-            expected_root_hash,
+            expected_root_hash.0,
         );
 
         Ok(())
     }
 
+    /// This function computes a new merkle path on split insertion (ie when inserting a new value creates
+    /// a key split).
+    ///
+    /// Add the correct siblings of the new merkle path by finding the splitting nibble and
+    /// adding the default leaves to the path
+    ///
+    /// To compute the number of default leaves we need to add, we need to:
+    /// - Compute the number of default leaves to separate the old leaf from the new leaf in the last nibble
+    /// - Compute the number of default leaves to traverse the common prefix
+    /// - Compute the number of default leaves remaining to select the former old leaf in the former last nibble
+    /// (this leaf becomes an internal node, hence the path needs to be fully specified)
+    fn compute_new_merkle_path_on_split<V: AsRef<[u8]>>(
+        mut self,
+        leaf_node: SparseMerkleLeafNode,
+        new_element_key: KeyHash,
+        new_element_value: V,
+    ) -> SparseMerkleProof<H> {
+        let new_key_path = NibblePath::new(new_element_key.0.to_vec());
+        let old_key_path = NibblePath::new(leaf_node.key_hash.0.to_vec());
+
+        // The verify_nonexistence check from before ensure that the common prefix nibbles_len is greater than the
+        // siblings len
+        let mut new_key_nibbles = new_key_path.nibbles();
+        let mut old_key_nibbles = old_key_path.nibbles();
+
+        let common_prefix_len = skip_common_prefix(&mut new_key_nibbles, &mut old_key_nibbles);
+
+        let num_siblings = self.siblings().len();
+
+        // We can safely unwrap these values as the check have been already performed in verify_nonexistence
+        let mut new_key_bits = new_key_nibbles.bits();
+        let mut old_key_bits = old_key_nibbles.bits();
+
+        let mut default_siblings_leaf_nibble = 0;
+
+        while new_key_bits.next() == old_key_bits.next() {
+            default_siblings_leaf_nibble += 1;
+        }
+
+        let num_default_siblings = (4 - (num_siblings % 4)) % 4 /* the number of default leaves we need to add to the previous root */
+                            + ((4*(common_prefix_len+1) - num_siblings) / 4 ) * 4 /* the number of default leaves we need to add to the path */
+                            + (default_siblings_leaf_nibble)
+            - 4;
```

Comment:
Tagging this for future reference, since this simplifies to:

```rust
let num_default_siblings = (-num_siblings % 4) + (4*common_prefix_len) - num_siblings + default_siblings_leaf_nibble
```

If the unreduced expression is here to help readers grok the intuition behind it, then I would add at least a few lines of context about how you're getting there.